### PR TITLE
Refactors the context restart code.

### PR DESCRIPTION
**Pull request by Markil3**
https://hub.jmonkeyengine.org/t/white-screen-on-context-restart/43866/28?u=markil3
Now most of the code that goes into initializing the renderer is now repeated every time. This is accomplished by moving most of the logic from the initContextFirstTime() methods into a new initContext(boolean) method. On the first boot, the code ultimately executes the same way. However, every context restart calls initContext(false) now as well. This will repeat all the code from before save for the creation of the GL and Renderer instances as well as input initialization.

At the moment, SimpleApplication now has a new default binding for the TAB key that restarts the context. This is to make it easier to test various examples, and I plan on removing it before we actually merge it.

This should solve issues #844 and #1445. This commit should also replace PR #1524.

### CODE REVIEW

File: jme3-core/src/main/java/com/jme3/app/SimpleApplication.java [hunk 1]
```diff
@@ -92,6 +93,8 @@ public void onAction(String name, boolean value, float tpf) {
                 if (stateManager.getState(StatsAppState.class) != null) {
                     stateManager.getState(StatsAppState.class).toggleStats();
                 }
+            }else if (name.equals(INPUT_RESTART_CONTEXT)){
```
Comment (Markil3) [inline jme3-core/src/main/java/com/jme3/app/SimpleApplication.java:96]: The new input binding is for any testing anyone wants to do. Just let me know when we are ready to merge, and I'll remove it and put it in TestContextRestart.

**Comment by stephengold** [comment]
I confirmed that this PR resolves issue #844, by running the test case from https://hub.jmonkeyengine.org/t/nifty-does-not-properly-display-gui-after-app-restart-when-gamma-correction-is-enabled/40213/19 and repeatedly pressing the 'T' and 'Tab' keys.

File: jme3-examples/src/main/java/jme3test/renderer/TestContextRestart.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import com.jme3.math.Quaternion;
+import com.jme3.renderer.RenderManager;
+import com.jme3.renderer.ViewPort;
+import com.jme3.scene.Geometry;
+import com.jme3.scene.control.AbstractControl;
+import com.jme3.scene.shape.Box;
+import com.jme3.system.AppSettings;
+
+/**
+ * Tests how gamma correction works after a context restart.
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/renderer/TestContextRestart.java:50]: Please document the new test. For instance, how does one determine whether it succeeded or failed? Can it be run from the test chooser (which skips `main()`)?

File: jme3-examples/src/main/java/jme3test/renderer/TestContextRestart.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package jme3test.renderer;
+
+import com.jme3.app.SimpleApplication;
+import com.jme3.input.KeyInput;
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/renderer/TestContextRestart.java:35]: Please delete the 3 unused imports.

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglContext.java [hunk 4]
```diff
@@ -174,7 +174,28 @@ protected int getNumSamplesToUse() {
         return samples;
     }
 
+    /**
+     * Reinitializes the relevent details of the context. For internal use only.
```
Comment (stephengold) [inline jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglContext.java:178]: Please correct the spelling "relevent" -> "relevant".

**Comment by stephengold** [comment]
I ran `TestContextRestart` with both jme3-lwjgl and jme3-lwjgl3, and noticed it behaves differently depending on the LWJGL version.

With jme3-lwjgl (LWJGL 2.9.3) pressing 'Tab' causes the box to disappear. With jme3-lwjgl3 (LWJGL 3.2.3 build 13) pressing 'Tab' causes a momentary flicker but the box does not disappear.

**Comment by stephengold** [comment]
@Markil3 do you intend to act on my comments?

**Comment by Markil3** [comment]
Things have gotten a bit crazy over on my end. I'll try to finish these by the weekend, though. Sorry! Completely forgot!

**Comment by Markil3** [comment]
So, I investigated your issue, and it seems to be linked to the AppSettings.setRenderer setting. I can change this to whatever I want on jme3-lwjgl3, but trying to set this on jme3-lwjgl seems to cause problems. I'm investigating it now, but at least it seems to work on the recommended library.

**Comment by Markil3** [comment]
Well, this is interesting: moving the call to initialize the GL renderer so that it only triggers upon the first run seems to solve the issue. However, trying to do the same on the jme3-lwjgl3 library breaks it if I'm using something other than OpenGL 2.0.

No idea what's going on underneath, but it works without any major problems, so I guess I'll just apply that fix.

File: jme3-lwjgl/src/main/java/com/jme3/system/lwjgl/LwjglContext.java [hunk 5]
```diff
@@ -251,42 +251,65 @@ protected int getNumSamplesToUse() {
         return samples;
     }
 
+    /**
+     * Reinitializes the relevent details of the context. For internal use only.
```
Comment (stephengold) [inline jme3-lwjgl/src/main/java/com/jme3/system/lwjgl/LwjglContext.java:255]: Please correct the spelling of "relevant". "Relevent" is not a word in English.

**Comment by stephengold** [comment]
In the interest of keeping things moving, I've decided to integrate this and fix the spelling later.

**Comment by stephengold** [comment]
@Markil3 Thank you for contributing to JMonkeyEngine!

**Comment by stephengold** [comment]
Verified the fix in v3.4.0-beta4.

**Pull request merged**